File: src/main/java/org/modelingvalue/nelumbo/logic/UniversalQuantifier.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.logic;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Set;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.InferContext;
import org.modelingvalue.nelumbo.InferResult;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Variable;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class UniversalQuantifier extends Quantifier {
    @Serial
    private static final long serialVersionUID = 8390514736994316431L;

    private static Functor    FUNCTOR;

    static {
        KnowledgeBase.registerFunctorSetter(UniversalQuantifier.class, f -> FUNCTOR = f);
    }

    public UniversalQuantifier(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args);
    }

    protected UniversalQuantifier(List<AstElement> elements, List<Variable> localVars, Predicate predicate) {
        super(FUNCTOR, elements, localVars, predicate);
    }

    private UniversalQuantifier(Object[] args, UniversalQuantifier declaration) {
        super(args, declaration);
    }

    @Override
    protected UniversalQuantifier struct(Object[] array, Node declaration) {
        return new UniversalQuantifier(array, (UniversalQuantifier) declaration);
    }

    @Override
    protected InferResult resolve(InferContext context, InferResult predResult) {
        List<Variable> localVars = localVars();
        Set<Predicate> facts = Set.of(), falsehoods = Set.of();
        boolean completeFacts = true, completeFalsehoods = true;
        for (Predicate predFalsehood : predResult.falsehoods()) {
            Predicate falsehood = setBinding(predFalsehood.getBinding().removeAllKey(localVars));
            if (falsehood.isFullyBound()) {
                falsehoods = falsehoods.add(falsehood);
            } else {
                completeFalsehoods = false;
            }
        }
        for (Predicate predFact : predResult.facts()) {
            Predicate fact = setBinding(predFact.getBinding().removeAllKey(localVars));
            if (!falsehoods.contains(fact)) {
                if (fact.isFullyBound()) {
                    facts = facts.add(fact);
                } else {
                    completeFacts = false;
                }
            }
        }
        if (!isFullyBound()) {
            if (!predResult.completeFacts()) {
                completeFacts = false;
            }
            if (!predResult.completeFalsehoods()) {
                completeFalsehoods = false;
            }
        } else if (falsehoods.isEmpty() && facts.isEmpty()) {
            facts = facts.add(this);
        }
        return InferResult.of(facts, completeFacts, falsehoods, completeFalsehoods, predResult.cycles());
    }

}

File: web-ts/src/logic/UniversalQuantifier.ts
/**
 * UniversalQuantifier - forall quantifier.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.UniversalQuantifier
 */

import { List, Set } from 'immutable';
import type { AstElement } from '../AstElement';
import { Variable } from '../Variable';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { Predicate } from './Predicate';
import { Quantifier } from './Quantifier';
import { InferResult } from '../InferResult';
import type { InferContext } from '../InferContext';

/**
 * UniversalQuantifier - "for all" quantification.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.UniversalQuantifier
 */
export class UniversalQuantifier extends Quantifier {
  // @JAVA_REF org.modelingvalue.nelumbo.logic.UniversalQuantifier#FUNCTOR
  static FUNCTOR: Functor | null = null;

  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, ...args);
  }

  protected static fromDataUniversal(data: unknown[], declaration?: Node): UniversalQuantifier {
    const uq = Object.create(UniversalQuantifier.prototype) as UniversalQuantifier;
    (uq as unknown as { _data: unknown[] })._data = data;
    (uq as unknown as { _declaration: Node })._declaration = declaration ?? uq;
    (uq as any)._binding = null;
    (uq as any)._hashCodeCached = false;
    (uq as any)._hashCode = 0;
    (uq as any)._nrOfUnbound = -1;
    return uq;
  }

  protected override struct(data: unknown[], declaration?: Node): UniversalQuantifier {
    return UniversalQuantifier.fromDataUniversal(data, declaration ?? this.declaration());
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.UniversalQuantifier#resolve(InferContext, InferResult)
  protected override resolveWithResult(_context: InferContext, predResult: InferResult): InferResult {
    const localVars = this.localVars();
    let facts = Set<Predicate>();
    let falsehoods = Set<Predicate>();
    let completeFacts = true;
    let completeFalsehoods = true;

    for (const predFalsehood of predResult.falsehoods()) {
      const binding = predFalsehood.getBinding();
      const cleanBinding = binding !== null ? binding.deleteAll(localVars) : binding;
      const falsehood = this.setBinding(cleanBinding!);
      if (falsehood.isFullyBound()) {
        falsehoods = falsehoods.add(falsehood);
      } else {
        completeFalsehoods = false;
      }
    }

    for (const predFact of predResult.facts()) {
      const binding = predFact.getBinding();
      const cleanBinding = binding !== null ? binding.deleteAll(localVars) : binding;
      const fact = this.setBinding(cleanBinding!);
      if (!falsehoods.contains(fact)) {
        if (fact.isFullyBound()) {
          facts = facts.add(fact);
        } else {
          completeFacts = false;
        }
      }
    }

    if (!this.isFullyBound()) {
      if (!predResult.completeFacts()) {
        completeFacts = false;
      }
      if (!predResult.completeFalsehoods()) {
        completeFalsehoods = false;
      }
    } else if (falsehoods.isEmpty() && facts.isEmpty()) {
      facts = facts.add(this as Predicate);
    }

    return InferResult.of(facts, completeFacts, falsehoods, completeFalsehoods, predResult.cycles());
  }
}
